server/functions/framework/compiled_function.go

```diff
@@ -193,232 +201,334 @@ func (c *CompiledFunction) Eval(ctx *sql.Context, row sql.Row) (interface{}, err
-			preferredCasts = append([][]TypeCastFunction{}, matchCasts[matchIdx])
+			preferredOverloads = append([]overloadMatch{}, cand)
 		} else if currentPreferredCount == preferredCount {
-			preferredOverloads = append(preferredOverloads, match)
-			preferredCasts = append(preferredCasts, matchCasts[matchIdx])
+			preferredOverloads = append(preferredOverloads, cand)
 		}
 	}
-	// Check once more for exactly one match
-	if len(preferredOverloads) == 1 {
-		matchedOverload := c.Functions
-		for _, parameter := range preferredOverloads[0] {
-			matchedOverload = matchedOverload.Parameter[parameter]
+	return preferredOverloads
+}
+
+// exactTypeMatches returns the set of overload candidates that have the most exact type matches for the arg types
+// provided.
+func exactTypeMatches(argTypes []pgtypes.DoltgresType, candidates []overloadMatch) []overloadMatch {
+	matchCount := 0
+	var matches []overloadMatch
+	for _, cand := range candidates {
+		currentMatchCount := 0
+		for argIdx := range argTypes {
+			// For variadic functions, the final param consumes all remaining arguments
+			var paramType pgtypes.DoltgresTypeBaseID
+			if argIdx >= len(cand.params.paramTypes) {
+				paramType = cand.params.paramTypes[len(cand.params.paramTypes)-1]
+			} else {
+				paramType = cand.params.paramTypes[argIdx]
+			}
+
+			// NULL values count as exact matches, since all types accept NULL as a valid value
+			paramBaseID := argTypes[argIdx].BaseID()
+			if paramBaseID == paramType || paramBaseID == pgtypes.DoltgresTypeBaseID_Null {
+				currentMatchCount++
+			}
+		}
+		if currentMatchCount > matchCount {
+			matchCount = currentMatchCount
+			matches = append([]overloadMatch{}, cand)
+		} else if currentMatchCount == matchCount {
+			matches = append(matches, cand)
+		}
+	}
+	return matches
+}
+
+// typeCompatibleOverloads returns all overloads that have a matching number of params whose types can be
+// implicitly converted to the ones provided. This is the set of all possible overloads that could be used with the
+// param types provided.
+func (c *CompiledFunction) typeCompatibleOverloads(argTypes []pgtypes.DoltgresType, sources []Source) []overloadMatch {
+	var compatible []overloadMatch
+	for _, overload := range c.ParamPermutations {
+		if !paramNumbersMatch(argTypes, overload) {
+			continue
+		}
+
+		isConvertible := true
+		overloadCasts := make([]TypeCastFunction, len(argTypes))
+		// Polymorphic parameters must be gathered so that we can later verify that they all have matching base types
+		var polymorphicParameters []pgtypes.DoltgresType
+		var polymorphicTargets []pgtypes.DoltgresType
+		for i := range argTypes {
+			// For variadic functions, the final param consumes all remaining arguments
+			var paramType pgtypes.DoltgresTypeBaseID
+			if i >= len(overload.paramTypes) {
+				paramType = overload.paramTypes[len(overload.paramTypes)-1]
+			} else {
+				paramType = overload.paramTypes[i]
+			}
```

Comment:
There is a far better way to do this, and wouldn't require these lines or _any_ changes in any of the code in the new subfunctions.

You've created `overloadParamPermutation`, so we can simply rename/extend that. Have it something like:
```go
// TODO: doc
type extendedParamPermutation struct {
	function      FunctionInterface
	paramTypes    []pgtypes.DoltgresTypeBaseID
	extendedTypes []pgtypes.DoltgresTypeBaseID
	variadic      int
}
```
Change `variadic` to an `int` so that `-1` can represent "not variadic", otherwise it's an index pointing into `paramTypes` stating which one is variadic. All of the new subfunctions would then work on `extendedTypes`, so none of them would know or care that the function is variadic. If there are no variadic parameters, then `paramTypes == extendedTypes`.

We can add a new function onto it as well, to convert the parameters that were given to the values that the variadic (and non-variadic) functions expect. We'd call this before passing the values to `Callable`.
```go
// coalesceVariadicValues returns a new value set that coalesces all variadic parameters into their actual array parameter.
func (epp *extendedParamPermutation) coalesceVariadicValues(values []any) []any {
	// If the overload is not variadic, then we don't need to do anything
	if epp.variadic < 0 {
		return values
	}
	coalescedValues := make([]any, len(epp.paramTypes))
	copy(coalescedValues, values[:epp.variadic])
	// This is for the values after the variadic index, so we need to add 1. We subtract the extended parameter count
	// from the actual parameter count to obtain the additional parameter count.
	firstValueAfterVariadic := epp.variadic+1+(len(epp.extendedTypes)-len(epp.paramTypes))
	copy(coalescedValues[epp.variadic+1:], values[firstValueAfterVariadic:])
	// We can just take the relevant slice out of the given values to represent our array, since all arrays use []any
	coalescedValues[epp.variadic] = values[epp.variadic:firstValueAfterVariadic]
	return coalescedValues
}
```

Next, we need to flatten the tree, so we can change `FunctionOverloadTree` into something like the following:
```go
// Overloads represents the collection of all valid overloads for a function.
type Overloads struct {
	// ExactMatches uses the concatenation of the parameters' base IDs to construct an exact lookup key for each
	// overload. This is only used by a specific step during function resolution, and therefore does not apply to other
	// steps (such as resolving the `unknown` type, etc.).
	ExactMatches map[string]FunctionInterface
	// Permutations contains all of the valid permutations for the function.
	Permutations []FunctionInterface
}

// Add adds the given function to the overload collection. Returns false if the new overload collides with an existing
// overload.
func (overloads *Overloads) Add(function FunctionInterface) bool {
	key := overloads.BaseIDsToKey(function.GetParameterBaseIDs())
	if _, ok := overloads.ExactMatches[key]; ok {
		return false
	}
	overloads.ExactMatches[key] = function
	return true
}

// BaseIDsToKey returns a string key that may be used in the `ExactMatches` field to find the desired overload.
func (overloads *Overloads) BaseIDsToKey(baseIDs []pgtypes.DoltgresTypeBaseID) string {
	// We'll use this buffer to create our key from the base IDs. Since each base ID is an uint32, we multiply the
	// length by 4.
	keyBuffer := utils.NewWriter(uint64(len(baseIDs)*4))
	for _, baseID := range baseIDs {
		keyBuffer.Uint32(uint32(baseID))
	}
	keyBytes := keyBuffer.Data()
	// Use the unsafe package so that we're reusing the underlying buffer to create our key
	return unsafe.String(unsafe.SliceData(keyBytes), len(keyBytes))
}

// expandParameters returns all of the permutations, while using the given parameter length to determine the length
// target for variadic functions.
func (overloads *Overloads) expandParameters(paramLength int) []extendedParamPermutation {
	extended := make([]extendedParamPermutation, len(overloads.Permutations))
	for permutationIdx, permutation := range overloads.Permutations {
		params := permutation.GetParameterBaseIDs()
		variadicIndex := permutation.VariadicIndex()
		if variadicIndex >= 0 && len(params) <= paramLength {
			// Variadic functions may only match when the function is declared with parameters that are fewer or equal
			// to our target length. If our target length is less, then we cannot expand, so we do not treat it as
			// variadic.
			extendedParams := make([]pgtypes.DoltgresTypeBaseID, paramLength)
			copy(extendedParams, params[:variadicIndex])
			// This is copying the parameters after the variadic index, so we need to add 1. We subtract the declared
			// parameter count from the target parameter count to obtain the additional parameter count.
			firstValueAfterVariadic := variadicIndex+1+(paramLength-len(params))
			copy(extendedParams[firstValueAfterVariadic:], params[variadicIndex+1:])
			// ToArrayType immediately followed by BaseType is a way to get the base type without having to cast.
			// For array types, ToArrayType causes them to return themselves.
			variadicBaseType := permutation.GetParameters()[variadicIndex].ToArrayType().BaseType().BaseID()
			for variadicParamIdx := 0; variadicParamIdx < 1+(paramLength-len(params)); variadicParamIdx++ {
				extendedParams[variadicParamIdx+variadicIndex] = variadicBaseType
			}
			extended[permutationIdx] = extendedParamPermutation{
				function:      permutation,
				paramTypes:    params,
				extendedTypes: params,
				variadic:      variadicIndex,
			}
		} else {
			extended[permutationIdx] = extendedParamPermutation{
				function:      permutation,
				paramTypes:    params,
				extendedTypes: params,
				variadic:      -1,
			}
		}
	}
	return extended
}
```
And then we just change `for _, overload := range c.ParamPermutations` to `for _, overload := range c.Overloads.expandParameters(len(argTypes))` and everything should more or less "just work". This also handles things like variadic parameters that are not at the end (which we'll eventually have to support, perhaps soon) and makes it relatively trivial to add support for specifically calling variadic functions: (see `variadic_example(VARIADIC array[0.0])` in the function resolution documentation: https://www.postgresql.org/docs/15/typeconv-func.html).